Chapter on Internationalization and Localization

igorsantos07 opened the pull request:
Inclusion of material regarding pure i18n implementation, related to Gettext and Poedit.

Refs #623.

igorsantos07 committed:
Adding basic explanation on i18n and section TODOs

- what's i18n, l10n and pluralization forms for
- common methods of implementation
- added the rest of the sub-sections, with TODOs

igorsantos07 committed:
i18n: Adding info on extensions, files and folders

igorsantos07 commented:
I'm currently working on with LingoHub to get their approval for quotations and reference linking, as their article is a great resource for this subject but is copyrighted. Thus, some sections (such as key naming) were skipped until a decision is made by them.

drbyte commented on _posts/05-06-01-Internationalization-and-Localization.md:
```diff
@@ -0,0 +1,133 @@
+without refactors. This is usually done once - preferably, in the beginning of the project, or else you'll probably
+need some huge changes in the source!
+- **Localization** happens when you adapt the interface (mainly) by translating contents, based on the i18n work done
+before. It usually us done every time a new language or region needs support, and is updated when new interface pieces
```
> "It usually us done"

Should be "is", not "us"

drbyte commented on _posts/05-06-01-Internationalization-and-Localization.md:
```diff
@@ -0,0 +1,133 @@
+are added, as they need to be available in all supported languages.
+- **Pluralization** defines the rules needed between different languages to interoperate strings containing numbers and 
+counters. For instance, in English when you have only one item, it's singular, and anything different from that is 
+called plural; plural is this language is indicated by adding an S after some words, and sometimes changes parts of it.
```
> ; plural is this language is

should be "plural _in_ this language is"

drbyte commented on _posts/05-06-01-Internationalization-and-Localization.md:
```diff
@@ -0,0 +1,133 @@
+There are some cases, in big projects, where you might need to separate translations when the same words convey 
+different meaning given a context. In those cases you split them into different _domains_. They're basically named
+groups of POT/PO/MO files, where the filename is the said _translation domain_. Small and medium-sized projects usually,
+for simplicity, use only one domain; it's name is arbitrary, but we will be using "main" for our code samples.
```
> it's name is arbitrary

In this case, `it's` should be `its`

igorsantos07 commented:
Thanks @drbyte, that's what happens when you write text after midnight :D
Fixed locally.

igorsantos07 committed:
i18n: typos, keys, plurals and samples

igorsantos07 committed:
i18n: review and finishing post

- part about key types has been reviewed
- included the pieces about using Poedit, caching issues and usual helper functions

petk commented on _posts/05-06-01-Internationalization-and-Localization.md:
```diff
@@ -0,0 +1,348 @@
+{% endhighlight %}
+
+#### 3. Preparing translation for the first run
+To make matters easier - and one of the powerful advantages Gettext has over custom framework i18n packages - is it's custom file type. "Oh man, that's quite hard to understand and edit by hand, a simple array would be easier!" Make no mistake, applications like [Poedit] are here to help - _a lot_. You can get the program from [their website], it's free and available for all platforms. It's a pre ...
```
responsability -> *responsibility

igorsantos07 committed:
i18n: fixing typos and right column margin

drbyte commented on _posts/05-06-01-Internationalization-and-Localization.md:
```diff
@@ -0,0 +1,348 @@
+1. _`msgid` as a real sentence_.  
+    The main advantage are: 
+    - if there's pieces of the software untranslated in any given language, the key displayed will still maintain some
+    meaning. Example: if you happen to translate by heart from English to Spanish but needs help to translate to French,
```
3 suggestions:

The main ~~advantage~~ `advantages` are: 
    - if ~~there's~~ `there are` pieces of the software untranslated in any given language, the key displayed will still maintain some
    meaning. Example: if you happen to translate by heart from English to Spanish but ~~needs~~ `need` help to translate to French,

drbyte commented on _posts/05-06-01-Internationalization-and-Localization.md:
```diff
@@ -0,0 +1,348 @@
+    needed as a basis for other translations. Example: the developer would ideally have an `en.po` file, that
+    translators would read to understand what to write in `fr.po` for instance.
+    - missing translations would display meaningless keys on screen (`TOP_MENU_WELCOME` instead of `Hello there, User!`
+    on the said untranslated French page). That's good as would force translation to be complete before publishing -
```
insert `it`:

That's good as `it` would force translation to be complete

drbyte commented on _posts/05-06-01-Internationalization-and-Localization.md:
```diff
@@ -0,0 +1,348 @@
+### Everyday usage
+In a common application, you would use some Gettext functions while writing static text in your pages. Those sentences
+would then appear in `.po` files, get translated, compiled into `.mo` files and then, used by Gettext when rendering
+the actual interface. Given that, let's tie together what we have discussed so far in a a step-by-step example:
```
double `a` `a`:

discussed so far in ~~a~~ a step-by-step example:

drbyte commented on _posts/05-06-01-Internationalization-and-Localization.md:
```diff
@@ -0,0 +1,348 @@
+                     $unread),
+            $unread)?>
+        </h2>
+    <? endif ?>
```
Given that you used `<?php` as your open tag earlier, I recommend also using it on your `<? endif ?>` line

drbyte commented on _posts/05-06-01-Internationalization-and-Localization.md:
```diff
@@ -0,0 +1,348 @@
+<p><?=gettext('We\'re now translating some strings')?></p>
+{% endhighlight %}
+
+- [`gettext()`][func] simply translates a `msgid` into it's corresponding `msgstr` for a given language. There's also
```
Remove the apostrophe in `it's` here:

> into it's corresponding

drbyte commented on _posts/05-06-01-Internationalization-and-Localization.md:
```diff
@@ -0,0 +1,348 @@
+    // if the cookie is present instead, let's just keep it
+    $lang = $_COOKIE['lang']; //you should sanitize this!
+} elseif (isset($_SERVER['HTTP_ACCEPT_LANGUAGE'])) {
+    // default resort: look for the languages the browser says the user accepts
```
>    // default resort: look for the languages the browser says the user accepts

I'd recommend removing the word "resort", as it's not needed and in this context is used awkwardly.

drbyte commented on _posts/05-06-01-Internationalization-and-Localization.md:
```diff
@@ -0,0 +1,348 @@
+{% endhighlight %}
+
+#### 3. Preparing translation for the first run
+To make matters easier - and one of the powerful advantages Gettext has over custom framework i18n packages - is it's custom file type. "Oh man, that's quite hard to understand and edit by hand, a simple array would be easier!" Make no mistake, applications like [Poedit] are here to help - _a lot_. You can get the program from [their website], it's free and available for all platforms. It's a pre ...
```
1. Another `it's` which should be `its`:
   
   > is it's custom file type
2. I suggest removing the last half of this sentence, as "using with responsibility the powers" is awkward wording, and I'm not sure this last phrase is needed to still convey the intended message:
   
   > It's a pretty easy tool to get used to, and a very powerful one at the same time~~ - using with responsibility the powers Gettext gave it~~.
3. The URL?
   
   > from [their website]

drbyte commented on _posts/05-06-01-Internationalization-and-Localization.md:
```diff
@@ -0,0 +1,348 @@
+- Project name and version, Translation Team and email address: useful information that goes in the `.po` file header;
+- Language: here you should use that format we mentioned before, such as `en_US` or `pt_BR`;
+- Charsets: UTF-8, preferably;
+- Source charset: set here the charset used by your PHP files - probably UTF-8 AS well, right?
```
> probably UTF-8 AS well

Should `AS well` be `as well` ?

drbyte commented on _posts/05-06-01-Internationalization-and-Localization.md:
```diff
@@ -0,0 +1,348 @@
+After setting those points you'll be prompted to save the file - using that directory structure we mentioned as well, and then it will run a scan through your source files to find the localization calls. They'll be fed empty into the translation table, and you'll start typing in the localized versions of those strings. Save it and a `.mo` file will be (re)compiled into the same folder and ta-dah: ...
+
+#### 4. Translating strings
+As you may have noticed before, there are two main types of localized strings: simple ones and the ones with plural forms. The first ones have simply two boxes: source and localized string. The source string can't be modified as Gettext/Poedit do not include the powers to alter your source files - you should change the source itself and rescan the files. Tip: you may right-click a translation lin ...
```
> with the source files and lines where that string is being.

I think there's more to this sentence... ?

Or maybe just change `being` to `found` ?

drbyte commented on _posts/05-06-01-Internationalization-and-Localization.md:
```diff
@@ -0,0 +1,348 @@
+#### Additional helper functions
+As preferred by many people, it's easier to use `_()` instead of `gettext()`. Many custom i18n libraries from frameworks use something similar to `t()` as well, to make translated code shorter. However, that's the only function that sports a shortcut. You might want to add in your project some others, such as `__()` or `_n()` for `ngettext()`, or maybe a fancy `_r()` that would join `gettext()` a ...
+
+In those cases, you'll need to instruct the Gettext utility on how to extract the strings from those new functions. Don't be afraid, it's something very easy. It's just a field in the `.po` file, or a Settings screen on Poedit - remember when we mentioned it before? In the editor that option is inside "Catalog > Properties > Source keywords". You need to include there the specifications of those  ...
```
1. Don't need "something" here:
   
   > Don't be afraid, it's ~~something~~ very easy.
   
   or even just:
   
   > It's very easy.
2. This "remember" phrase reflects a different tone from the rest of the document.
   
   > It's just a field in the `.po` file, or a Settings screen on Poedit ~~remember when we mentioned it before?~~

drbyte commented on _posts/05-06-01-Internationalization-and-Localization.md:
```diff
@@ -0,0 +1,348 @@
+- if you create something like `t()` that simply returns the translation for a string, you can specify it as `t`. Gettext will know the only function argument is the string to be translated;
+- if the function has more than one argument, you can specify in which one the first string is - and if needed, the plural form as well. For instance, if we call our function like this: `__('one user', '%d users', $number)`, the specification would be `__:1,2`, meaning the first form is the first argument, and the second form is the second argument. If your number comes as first argument instead, ...
+
+After including those new rules in the `.po` file, a new scan will bring your new strings just as easy as before.
```
> will `bring` your new strings

When you say `bring`, do you mean `bring in`?  Or would it be more specific to say `index`?

igorsantos07 commented:
@drbyte thanks for all that effort on reviewing :)
Besides not being native, I've been writing that during late night, not the most attentive part of my day :sweat_smile:

drbyte commented:
> Besides not being native, I've been writing that during late night, not the most attentive part of my day 

Hey, for “not being native”, your command of English is good! :)

philsturgeon commented:
@drbyte You speak better English than I speak Portugese that's for sure. I've been trying to learn Spanish for a year and I'm just about good enough to get through a meal, and have a few useless bastard phrases boiled into my head like "No es un huevo, es un manzana!" 

I have much respect for anyone contributing in a language other than their first. Hopefully you weren't be embarrassed somebody pointed out the typos and tweaks, it's all part of the process. :)

drbyte commented:
Agreed @philsturgeon . I've huge respect for what @igorsantos07 has contributed, particularly writing in non-native language.

And, I've learned some stuff about `gettext` that I didn't know before too :)

igorsantos07 commented:
@drbyte that's for sure what makes me happy :D

Regarding get embarrassed, those where mostly typos and result of a lack of attention, at most. I would get if I had written uninteligible babble hahahaha

igorsantos07 commented:
@codeguy any news on this? :)

igorsantos07 commented:
@codeguy was this forgotten? :(

drbyte commented:
When I last reviewed it, my opinion was 👍 
Hopefully it will get merged.

philsturgeon commented:
It's got a +1 from me. I threw this out on Twitter to see if there is any further feedback, and if nothing has happened in 24 hours I'll click that shiny green button.

heiglandreas commented on _posts/05-06-01-Internationalization-and-Localization.md:
```diff
@@ -0,0 +1,384 @@
+is called _full-stack_ for a reason, right?
+
+However, the most classic way and often taken as reference for i18n and l10n is a [Unix tool called `gettext`][gettext].
+It dates back to 1995 and is still the most complete implementation for translating software. It is pretty easy to get
```
Personally I'd favour mentioning of other tools here that might be used by professional translators (which, as far as I know, gettext usually isn't) for example [XLIFF](https://en.wikipedia.org/wiki/XLIFF).

igorsantos07 replied:
Uh-oh. It seems I had forgotten the suggestions here, made by @heiglandreas and @localheinz. Worked on those already!

codeguy commented:
+1 from me. @philsturgeon if this is ready to merge with no more feedback, go ahead and merge.

philsturgeon commented on _posts/05-06-01-Internationalization-and-Localization.md:
```diff
@@ -0,0 +1,384 @@
+is called _full-stack_ for a reason, right?
+
+However, the most classic way and often taken as reference for i18n and l10n is a [Unix tool called `gettext`][gettext].
+It dates back to 1995 and is still the most complete implementation for translating software. It is pretty easy to get
```
I'm always a fan to add more relevant tools and alternatives in, even if they're not the _most_ popular. Maybe if theres a Composer-based one floating around that'll help those unable to install extensions and the like too.

localheinz replied:
@philsturgeon 

How about 
- https://github.com/oscarotero/Gettext
- https://github.com/symfony/translation
- https://github.com/zendframework/zend-i18n

?

philsturgeon replied:
Sure!

igorsantos07 replied:
ok, just to make sure I got it right. What's missing is citing some sort of library like this (awesomely giant) Gettext?
I liked the idea, it's a great addition :)

philsturgeon replied:
Yeah, generally we don't just have one tool, we put a few decent options in so people can find the right thing. So a user and gettext and maybe a popular alternative would be two good additions to what you have. If you know of any! 

## 

Phil Sturgeon
Sent from my iPhone and there's probably typos because I'm probably at the pub.

> On Jul 20, 2016, at 16:29, Igor Santos notifications@github.com wrote:
> 
> In _posts/05-06-01-Internationalization-and-Localization.md:
> 
> > +languages with a total of four, five or six forms, such as Slovenian, Irish or Arabic.
> > +
> > +## Common ways to implement
> > +The easiest way to internationalize PHP software is by using array files and using those strings in templates, such as
> > +`<h1><?=$TRANS['title_about_page']?></h1>`. This is, however, hardly a recommended way for serious projects, as it poses
> > +some maintenance issues along the road - some might appear in the very beginning, such as pluralization. So, please,
> > +don't try this if your project will contain more than a couple of pages.
> > +
> > +Some frameworks will sport their own i18n packages. Those usually are a more powerful version of the above approach,
> > +but including features needed for real localization, such as plural forms and string replacement. You're free to use
> > +those if you feel like, but you might find bothering to edit array source files, having to deal with pure code issues
> > +(such as string scaping and so on). The main pro here is integration with the environment you're using - the framework
> > +is called _full-stack_ for a reason, right?
> > +
> > +However, the most classic way and often taken as reference for i18n and l10n is a [Unix tool called `gettext`][gettext].
> > +It dates back to 1995 and is still the most complete implementation for translating software. It is pretty easy to get
> > ok, just to make sure I got it right. What's missing is citing some sort of library like this (awesomely giant) Gettext?
> > I liked the idea, it's a great addition :)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

localheinz replied:
@igorsantos07 @philsturgeon 

Updated the list above!

philsturgeon replied:
Cool, work that into the PR and I'll get merging. :)

igorsantos07 committed:
Adding information on other i18n libraries

Cleaning up framework references, as they're listed below

igorsantos07 committed:
Passing the i18n text against the [free Grammarly checker](http://gram.ly/tYyH)

igorsantos07 commented:
@codeguy, @philsturgeon: I added the forgotten suggestions and also [checked the text](http://gram.ly/tYyH) for a couple of more subtle grammar mistakes :)

I _think_ it's now finally ready? Almost a year taking dust, sorry for making this even longer.

philsturgeon commented:
Great work @igorsantos07! Thank you for following through on the requested changes. FOSS is not always speedy, but the willingness to keep at it is appreciated.
